File: BOJ/5052.cpp
#include <stdio.h>
#include <string.h>
#include <algorithm>
using namespace std;
typedef long long ll;
struct Trie {
	Trie* tree[10];
	bool finish;
	bool cur;
	Trie() {
		fill(tree, tree + 10, nullptr);
		finish = false,cur=false;
	}
	~Trie() {
		for (int i = 0; i < 10; i++) if (tree[i]) delete tree[i];
	}

	void insert(char *key) {
		if (*key == NULL) {
			finish = true;
			return;
		}
		else {
			int next = *key - '0';
			if (!tree[next]) tree[next] = new Trie;
			cur = true;
			tree[next]->insert(key + 1);
		}
	}

	bool check() {
		if (finish && cur) return true;
		for (int i = 0; i < 10; i++) {
			if (tree[i]) {
				if (tree[i]->check()) return true;
			}
		}
		return false;
	}
};
char s[10001][11];
int main() {
	int t;
	scanf("%d", &t);
	while (t--) {
		int n;
		Trie* trie = new Trie;
		scanf("%d", &n);
		for (int i = 0; i < n; i++) {
			scanf("%s", s[i]);
			trie->insert(s[i]);
		}
		int no = 0;
		if (trie->check()) printf("NO\n");
		else printf("YES\n");
		delete trie;
	}
	return 0;
}

File: BOJ(Java 공부용)/5052.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int t =Integer.parseInt(st.nextToken());
        while(t-- >0) {
            Trie trie = new Trie();
            st = new StringTokenizer(br.readLine());
            int n =Integer.parseInt(st.nextToken());
            for(int i=0;i<n;i++) {
                String s = br.readLine();
                insert(s,trie);
            }
            if(check(trie)) bw.write("NO\n");
            else bw.write("YES\n");
        }
        bw.flush();
        bw.close();
    }
    private static class Trie {
        private Trie[] letter;
        private boolean cur;
        private boolean finish;
        private Trie() {
            this.cur = false;
            this.letter = new Trie[10];
            this.finish = false;
        }
    }
    private static boolean check(Trie trie) {
        if(trie.finish && trie.cur ) return true;
        for(int i=0;i<10;i++) {
            if(trie.letter[i] == null) continue;
            if(check(trie.letter[i])) {
                return true;
            }
        }
        return false;
    }
    private static void insert(String s,Trie trie) {
        for(int i=0;i<s.length();i++) {
            int c = s.charAt(i)-'0';
            if(trie.letter[c] == null) trie.letter[c] = new Trie();
            trie.cur = true;
            trie = trie.letter[c];
        }
        trie.finish = true;
    }
}